spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,223 @@
+    List<CombinedScanTask> runningTasks = Lists.newArrayList();
+    for (SnapshotTasks s : snapshotTasksList) {
+      if (s.snapshotId() == end.snapshotId()) {
+        runningTasks.addAll(s.tasks().subList(0, end.index() + 1));
+        break;
+      } else {
+        runningTasks.addAll(s.tasks());
+      }
+    }
+
+    return runningTasks;
+  }
+
+  private StreamingOffset calculateInitialOffset() {
+    Snapshot current = table.currentSnapshot();
+    List<Long> snapshotIds = SnapshotUtil.ancestorIds(current, table::snapshot);
+    if (snapshotIds.isEmpty()) {
+      return null;
+    } else {
+      long initSnapshotId = snapshotIds.get(snapshotIds.size() - 1);
+      return new StreamingOffset(initSnapshotId, -1, true /* is_starting */);
+    }
+  }
+
+  private StreamingOffset calculateEndOffset(StreamingOffset start) {
+    List<SnapshotTasks> snapshotTasksList = getChangesWithRateLimit(start.snapshotId(),
+        start.index(), start.isStartingSnapshotId());
+    SnapshotTasks last = snapshotTasksList.isEmpty() ? null : snapshotTasksList.get(snapshotTasksList.size() - 1);
+
+    return last == null ? start : new StreamingOffset(last.snapshotId, last.size() - 1,
+        last.snapshotId() == start.snapshotId() /* is_starting */);
+  }
+
+  @VisibleForTesting
+  List<SnapshotTasks> getChanges(long snapshotId, int index, boolean isStarting) {
+    List<SnapshotTasks> snapshotTasksList = Lists.newArrayList();
+    long currentSnapshotId = table.currentSnapshot().snapshotId();
+
+    if (isStarting) {
+      CloseableIterable<CombinedScanTask> iter = buildTableScan().useSnapshot(snapshotId).planTasks();
+      SnapshotTasks snapshotTasks = new SnapshotTasks(snapshotId, Lists.newArrayList(iter));
+      snapshotTasksList.add(snapshotTasks);
+    }
+
+    Long startSnapshotId = isStarting ? Long.valueOf(snapshotId) : table.snapshot(snapshotId).parentId();
+    Preconditions.checkState(startSnapshotId != null, "Start snapshot id should be existed");
+    List<Long> snapshotIds = SnapshotUtil.snapshotIdsBetween(table, startSnapshotId, currentSnapshotId);
+    snapshotIds.add(startSnapshotId);
+
+    for (int i = snapshotIds.size() - 1; i > 0; i--) {
+      long fromSnapshotId = snapshotIds.get(i);
+      long toSnapshotId = snapshotIds.get(i - 1);
+      CloseableIterable<CombinedScanTask> iter = buildTableScan()
+          .appendsBetween(fromSnapshotId, toSnapshotId)
+          .planTasks();
+      SnapshotTasks snapshotTasks = new SnapshotTasks(toSnapshotId, Lists.newArrayList(iter));
+      snapshotTasksList.add(snapshotTasks);
+    }
+
+    return snapshotTasksList.stream()
+        .map(t -> t.snapshotId() == snapshotId ?
+            new SnapshotTasks(snapshotId, t.tasks().subList(index + 1, t.size())) : t)
+        .filter(t -> t.size() > 0)
+        .collect(Collectors.toList());
+  }
+
+  @VisibleForTesting
+  List<SnapshotTasks> getChangesWithRateLimit(long snapshotId, int index, boolean isStarting) {
```

Comment:
I think I would prefer if this were orthogonal to `getChanges` instead of wrapping `getChanges`. I would expect this instead to accept all tasks and return a set of tasks for the rate limit:

```java
  static List<SnapshotTasks> rateLimit(List<SnapshotTasks> pendingTasks, int maxTasks) {
    ...
  }
```

Reply:
Done.